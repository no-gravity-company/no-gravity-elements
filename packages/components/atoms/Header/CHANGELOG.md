# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.2.11](https://github.com/no-gravity-company/no-gravity-elements/compare/@no-gravity-elements/header@1.2.10...@no-gravity-elements/header@1.2.11) (2024-12-27)

### Bug Fixes

- test if pipeline publish index.js for header ([daef339](https://github.com/no-gravity-company/no-gravity-elements/commit/daef339ac0af843d0cfa55375bc1ef7f92f2c2e1))

## [1.2.10](https://github.com/no-gravity-company/no-gravity-elements/compare/@no-gravity-elements/header@1.2.9...@no-gravity-elements/header@1.2.10) (2024-12-27)

### Bug Fixes

- add snapshot ([045997f](https://github.com/no-gravity-company/no-gravity-elements/commit/045997f61c7b69e7a31fcf551c850fa2edd8651a))
- show mobile with JS logic ([c2dbf42](https://github.com/no-gravity-company/no-gravity-elements/commit/c2dbf4270a1aca118bfa7cceb07587321895b999))

## [1.2.9](https://github.com/no-gravity-company/no-gravity-elements/compare/@no-gravity-elements/header@1.2.8...@no-gravity-elements/header@1.2.9) (2024-12-25)

### Bug Fixes

- add snapshot ([6699b4b](https://github.com/no-gravity-company/no-gravity-elements/commit/6699b4b5aeb8533f5610730d9cc1e2457d6fba5d))
- improve header and footer ([5fb0ee6](https://github.com/no-gravity-company/no-gravity-elements/commit/5fb0ee675d62d7f6c0e9d9c56749215137f381a1))

## [1.2.8](https://github.com/no-gravity-company/no-gravity-elements/compare/@no-gravity-elements/header@1.2.7...@no-gravity-elements/header@1.2.8) (2024-12-25)

### Bug Fixes

- use max-width + responsiveness ([9c828d6](https://github.com/no-gravity-company/no-gravity-elements/commit/9c828d6486f4211cf6660586176eb7bcfc5039e4))

## [1.2.7](https://github.com/no-gravity-company/no-gravity-elements/compare/@no-gravity-elements/header@1.2.6...@no-gravity-elements/header@1.2.7) (2024-12-16)

### Bug Fixes

- add transition css variable to header ([b149a51](https://github.com/no-gravity-company/no-gravity-elements/commit/b149a51be4b81edad515125e2c090eee840f893f))

## [1.2.6](https://github.com/no-gravity-company/no-gravity-elements/compare/@no-gravity-elements/header@1.2.5...@no-gravity-elements/header@1.2.6) (2024-12-16)

### Bug Fixes

- remove header margin right ([c02adfe](https://github.com/no-gravity-company/no-gravity-elements/commit/c02adfebde5da1bb3573c2d0a20a388fcafa929d))

## [1.2.5](https://github.com/no-gravity-company/no-gravity-elements/compare/@no-gravity-elements/header@1.2.4...@no-gravity-elements/header@1.2.5) (2024-12-13)

### Bug Fixes

- use event to change menu status ([c145f40](https://github.com/no-gravity-company/no-gravity-elements/commit/c145f405845b1801523af7ac43b05455a349743b))

## [1.2.4](https://github.com/no-gravity-company/no-gravity-elements/compare/@no-gravity-elements/header@1.2.3...@no-gravity-elements/header@1.2.4) (2024-12-13)

### Bug Fixes

- remove event after using it ([004989b](https://github.com/no-gravity-company/no-gravity-elements/commit/004989b5b19a1844d82c371539b3c5a1ea54c38a))

## [1.2.3](https://github.com/no-gravity-company/no-gravity-elements/compare/@no-gravity-elements/header@1.2.2...@no-gravity-elements/header@1.2.3) (2024-12-12)

### Bug Fixes

- add messageBuss to header WC ([31c9380](https://github.com/no-gravity-company/no-gravity-elements/commit/31c938082e19f36a63da2f722870e7a6d937604f))
- reove icon import from button because of unit tests ([c61fcd3](https://github.com/no-gravity-company/no-gravity-elements/commit/c61fcd3b0f944e37230aacdbc7f3350921a04a1f))

## [1.2.2](https://github.com/no-gravity-company/no-gravity-elements/compare/@no-gravity-elements/header@1.2.1...@no-gravity-elements/header@1.2.2) (2024-12-12)

### Bug Fixes

- avoid header menu icon size grow ([2efb2a6](https://github.com/no-gravity-company/no-gravity-elements/commit/2efb2a60f36894c0b6186b991532529d15ed0d02))

## [1.2.1](https://github.com/no-gravity-company/no-gravity-elements/compare/@no-gravity-elements/header@1.2.0...@no-gravity-elements/header@1.2.1) (2024-12-11)

### Bug Fixes

- add 100% width for footer and header ([778a8e0](https://github.com/no-gravity-company/no-gravity-elements/commit/778a8e0ac82cbdd608061785b7e5a5acb7502647))

# [1.2.0](https://github.com/no-gravity-company/no-gravity-elements/compare/@no-gravity-elements/header@1.1.2...@no-gravity-elements/header@1.2.0) (2024-12-11)

### Features

- add jsdoc for css variables on header and button ([2e07757](https://github.com/no-gravity-company/no-gravity-elements/commit/2e0775732333ff0eeebdb4ad490c19e4d4b29462))

## [1.1.2](https://github.com/no-gravity-company/no-gravity-elements/compare/@no-gravity-elements/header@1.1.1...@no-gravity-elements/header@1.1.2) (2024-12-10)

### Bug Fixes

- header shows options on mobile ([5ff75eb](https://github.com/no-gravity-company/no-gravity-elements/commit/5ff75ebdbe34ff189916ecf65d35321a116491c8))

## [1.1.1](https://github.com/no-gravity-company/no-gravity-elements/compare/@no-gravity-elements/header@1.1.0...@no-gravity-elements/header@1.1.1) (2024-12-10)

### Bug Fixes

- add css variables to nge-header ([1bb5d2a](https://github.com/no-gravity-company/no-gravity-elements/commit/1bb5d2a21bbe2a6c3982a6d8b0a014ecd3f7554b))

# 1.1.0 (2024-12-10)

### Features

- add header to atoms in storybook ([3d6ab04](https://github.com/no-gravity-company/no-gravity-elements/commit/3d6ab04248a2f69f7e36fb3857d8bd6206b40277))
- add header WC ([b8ade85](https://github.com/no-gravity-company/no-gravity-elements/commit/b8ade85c6ba761d5c87bb91ae058db8ccfcc2173))
- add value as observedAttributes ([9d183ae](https://github.com/no-gravity-company/no-gravity-elements/commit/9d183aeaba22f4396d98f1167418a17815b86286))
