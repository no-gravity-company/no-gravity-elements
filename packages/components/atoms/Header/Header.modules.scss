@import '../../../stylesheets/base/variables';

$header-height: var(--nge-header-height, baseUnit(8));
$header-padding: var(--nge-header-padding, baseUnit(2));
:host {
    ::slotted([slot="navigation"]) {
        display: flex;
        flex-direction: column;
        gap: baseUnit(1);      
    } 
    @media (min-width: 768px) {
        ::slotted([slot="navigation"]) {
            flex-direction: row;
        } 
    }
}

.nge-header {
  display: flex;
  align-items: center;
  padding: $header-padding;
  background-color: #fff;
  border-bottom: 1px solid #ddd;
  height: $header-height;
  box-sizing: border-box;
  gap: baseUnit(1);

  .logo {
    flex: 1;
  }

  .navigation {
    display: flex;
    gap: baseUnit(1);
  }

  .options {
    margin-left: auto;
  }

  .burger-menu {
    background: none;
    border: none;
    cursor: pointer;
    padding: 0;
    svg {
        transition: transform 500ms cubic-bezier(.4,0,.2,1), width 500ms cubic-bezier(.4,0,.2,1), height 500ms cubic-bezier(.4,0,.2,1), margin 500ms cubic-bezier(.4,0,.2,1);
        margin-right: calc(#{$header-padding} - calc(#{baseUnit(2)} * 2))
      }
      &.active svg {
        transform: rotate(90deg);
      }
      path {
        transition: transform 500ms cubic-bezier(0.4, 0, 0.2, 1),
          stroke-dasharray 500ms cubic-bezier(0.4, 0, 0.2, 1),
          stroke-dashoffset 500ms cubic-bezier(0.4, 0, 0.2, 1);
      }
      path:nth-child(1) {
        transform-origin: 36% 40%;
      }
      path:nth-child(2) {
        stroke-dasharray: 29 299;
      }
      path:nth-child(3) {
        transform-origin: 35% 63%;
      }
      path:nth-child(4) {
        stroke-dasharray: 29 299;
      }
      path:nth-child(5) {
        transform-origin: 61% 52%;
      }
      path:nth-child(6) {
        transform-origin: 62% 52%;
      }
      &.active path:nth-child(1) {
        transform: translateX(9px) translateY(1px) rotate(45deg);
      }
      &.active path:nth-child(2) {
        stroke-dasharray: 225 299;
        stroke-dashoffset: -72px;
      }
      &.active path:nth-child(3) {
        transform: translateX(9px) translateY(1px) rotate(-45deg);
      }
      &.active path:nth-child(4) {
        stroke-dasharray: 225 299;
        stroke-dashoffset: -72px;
      }
      &.active path:nth-child(5) {
        transform: translateX(9px) translateY(1px) rotate(-45deg);
      }
      &.active path:nth-child(6) {
        transform: translateX(9px) translateY(1px) rotate(45deg);
      }
      &.active svg {
        margin-right: calc(#{$header-padding} - calc(#{baseUnit(1.5)} * 2))
      }
  }

  .side-menu {
    position: fixed;
    top: $header-height;
    right: 0;
    width: 100%;
    box-sizing: border-box;
    height: 100%;
    background-color: #fff;
    box-shadow: 0px baseUnit(0.5) baseUnit(0.5) rgba(0, 0, 0, 0.5);
    display: flex;
    flex-direction: column;
    padding: $header-padding;
    z-index: 1000;
    transform: translateX(101%);
    transition: transform 0.3s ease;

    &.open {
      transform: translateX(0);
    }

    .close-button {
      align-self: flex-end;
      background: none;
      border: none;
      cursor: pointer;
      margin-bottom: baseUnit(1);
    }

    .mobile-nav {
      display: flex;
      flex-direction: column;
      gap: baseUnit(1);
    }
  }
}

@media (min-width: 768px) {
  .nge-header {
    .burger-menu {
      display: none;
    }

    .side-menu {
      display: none;
    }
  }
}
