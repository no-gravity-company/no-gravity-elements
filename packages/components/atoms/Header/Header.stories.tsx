import { Meta, StoryObj } from '@storybook/web-components';
import { html } from 'lit-html';

import '@no-gravity-elements/header';

const meta: Meta = {
  title: 'Header',
  component: 'nge-header',
  parameters: {
    webComponents: {
      tagName: 'nge-header',
    },
  },
  argTypes: {
  },
  render: () => html`
  <style type="text/css">
    .sb-show-main.sb-main-padded {
    padding: 0px;
    }
  </style>
  <nge-header>
    <div slot="logo">Logo</div>
    <nav slot="navigation">
      <div>Link 1</div>
      <div>Link 2</div>
      <div>Link 3</div>
    </nav>
    <div slot="options">User Options</div>
  </nge-header>`
};

export default meta;
type Story = StoryObj;

export const Default: Story = {
  args: {},
};
