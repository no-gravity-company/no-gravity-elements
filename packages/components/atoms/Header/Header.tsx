import { Fragment, FunctionComponent, h } from 'preact';
import { useState, useEffect } from 'preact/hooks';
import classNames from 'classnames';
import { MessageBusEvents } from '@no-gravity-elements/types';
import { messageBus } from '@no-gravity-elements/message-bus';
import './Header.modules.scss';

/**
 * <nge-header>
 *
 * Header component
 *
 * @element nge-header
 *
 * @slot logo - Slot for the logo content
 * @slot navigation - Slot for navigation links
 * @slot options - Slot for additional options
 *
 * @cssproperty --nge-header-background-color - Background color of the header.
 * @cssproperty --text-color - Text color of the header.
 * @cssproperty --nge-header-padding - Padding around the header content.
 * @cssproperty --nge-header-gap - Gap between header sections.
 * @cssproperty --nge-header-height - Height of the header.
 * @cssproperty --nge-border-bottom - Border bottom for header.
 * @cssproperty --nge-header-border-radius - Border radius of the header content.
 * @cssproperty --nge-header-icon-stroke - Color of the menu icon when header is on mobile
 * @cssproperty --nge-header-transition - Transition for header
 * 
 * @example
 * <nge-header>
 *   <div slot="logo">Logo</div>
 *   <div slot="navigation">Navigation Links</div>
 *   <div slot="options">User Options</div>
 * </nge-header>
 */

const Header: FunctionComponent = () => {
  const [isMobile, setIsMobile] = useState(window.innerWidth < 768);
  const [isMenuOpen, setIsMenuOpen] = useState(false);

  useEffect(() => {
    const handleResize = () => setIsMobile(window.innerWidth < 768);
    window.addEventListener('resize', handleResize);
    const eventId = messageBus.subscribe(MessageBusEvents.HEADER_MENU_ICON, (data) => {
      setIsMenuOpen(data.show);
    })
    return () => {
      window.removeEventListener('resize', handleResize);
      messageBus.unsubscribe(MessageBusEvents.HEADER_MENU_ICON, eventId)
    }
  }, []);

  const toggleMenu = () => {
    messageBus.publish(MessageBusEvents.HEADER_MENU_ICON, { show: !isMenuOpen })
  };

  return (
    <Fragment>
      <div className='nge-header'>
        <div className='logo'>
          <slot name='logo' />
        </div>

        {!isMobile && (
          <nav className='navigation'>
            <slot name='navigation' />
          </nav>
        )}

        <div className='options'>
          <slot name='options' />
        </div>

        {isMobile && (
          <button
            className={classNames('burger-menu', { active: isMenuOpen })}
            onClick={toggleMenu}
            aria-label='Toggle menu'
          >
            <svg
              xmlns='http://www.w3.org/2000/svg'
              width='40'
              height='40'
              viewBox='0 0 200 200'
            >
              <g stroke-width='6.5' stroke-linecap='round'>
                <path d='M72 82.286h28.75' fill='#000000' fill-rule='evenodd' stroke='#000000' />
                <path
                  d='M100.75 103.714l72.482-.143c.043 39.398-32.284 71.434-72.16 71.434-39.878 0-72.204-32.036-72.204-71.554'
                  fill='none'
                  stroke='#000000'
                />
                <path d='M72 125.143h28.75' fill='#000000' fill-rule='evenodd' stroke='#000000' />
                <path
                  d='M100.75 103.714l-71.908-.143c.026-39.638 32.352-71.674 72.23-71.674 39.876 0 72.203 32.036 72.203 71.554'
                  fill='none'
                  stroke='#000000'
                />
                <path
                  d='M100.75 82.286h28.75'
                  fill='#000000'
                  fill-rule='evenodd'
                  stroke='#000000'
                />
                <path
                  d='M100.75 125.143h28.75'
                  fill='#009100'
                  fill-rule='evenodd'
                  stroke='#000000'
                />
              </g>
            </svg>
          </button>
        )}

        {isMobile && (
          <aside className={classNames('side-menu', { open: isMenuOpen })}>
            <div className='mobile-nav'>
              <slot name='navigation' />
            </div>
          </aside>
        )}
      </div>
    </Fragment>
  );
};

export default Header;
