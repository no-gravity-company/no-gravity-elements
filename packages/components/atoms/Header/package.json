{
  "name": "@no-gravity-elements/header",
  "version": "1.2.11",
  "description": "No Gravity Element",
  "author": "No Gravity Team",
  "private": false,
  "homepage": "https://github.com/no-gravity-company/no-gravity-elements#readme",
  "license": "ISC",
  "main": "lib/index.js",
  "typings": "lib/index.d.ts",
  "directories": {
    "lib": "lib"
  },
  "files": [
    "lib/index.js",
    "lib/index.js.map",
    "types/index.d.ts"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/no-gravity-company/no-gravity-elements.git"
  },
  "scripts": {
    "build": "tsc --jsx react --emitDeclarationOnly --esModuleInterop --module ESNext --target ESNext --moduleResolution node --declaration --declarationDir ./lib/types src/*.ts && esbuild src/index.ts --bundle --outfile=lib/index.js --platform=node --target=es2018 --minify --sourcemap --external:react --external:react-dom",
    "publish-component": "yarn publish --access=public"
  },
  "bugs": {
    "url": "https://github.com/no-gravity-company/no-gravity-elements/issues"
  }
}
