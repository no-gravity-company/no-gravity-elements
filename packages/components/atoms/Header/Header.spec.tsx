import { h } from 'preact';
import { shallow } from 'enzyme';

import Header from './Header';

describe('Header', () => {
  it('should match the snapshot', () => {
    const wrapper = shallow(<Header>Test</Header>);
    expect(wrapper).toMatchSnapshot();
  });
});