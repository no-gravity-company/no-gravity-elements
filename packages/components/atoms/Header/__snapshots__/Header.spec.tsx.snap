// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should match the snapshot 1`] = `
<div
  className="nge-header"
>
  <div
    className="logo"
  >
    <slot
      name="logo"
    />
  </div>
  <nav
    className="navigation desktop-nav"
  >
    <slot
      name="navigation"
    />
  </nav>
  <div
    className="options"
  >
    <slot
      name="options"
    />
  </div>
  <button
    aria-label="Toggle menu"
    className="burger-menu"
    onClick={[Function]}
  >
    <svg
      height="40"
      viewBox="0 0 200 200"
      width="40"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        stroke-linecap="round"
        stroke-width="6.5"
      >
        <path
          d="M72 82.286h28.75"
          fill="#000000"
          fill-rule="evenodd"
          stroke="#000000"
        />
        <path
          d="M100.75 103.714l72.482-.143c.043 39.398-32.284 71.434-72.16 71.434-39.878 0-72.204-32.036-72.204-71.554"
          fill="none"
          stroke="#000000"
        />
        <path
          d="M72 125.143h28.75"
          fill="#000000"
          fill-rule="evenodd"
          stroke="#000000"
        />
        <path
          d="M100.75 103.714l-71.908-.143c.026-39.638 32.352-71.674 72.23-71.674 39.876 0 72.203 32.036 72.203 71.554"
          fill="none"
          stroke="#000000"
        />
        <path
          d="M100.75 82.286h28.75"
          fill="#000000"
          fill-rule="evenodd"
          stroke="#000000"
        />
        <path
          d="M100.75 125.143h28.75"
          fill="#009100"
          fill-rule="evenodd"
          stroke="#000000"
        />
      </g>
    </svg>
  </button>
  <aside
    className="side-menu"
  >
    <div
      className="mobile-nav"
    >
      <slot
        name="navigation"
      />
    </div>
  </aside>
</div>
`;
